Mathlib/Data/List/SplitBy.lean

```diff
@@ -139,6 +140,89 @@ theorem chain'_getLast_head_splitBy (r : α → α → Bool) (l : List α) :
     apply chain'_getLast_head_splitByLoop _ (not_mem_nil _) chain'_nil
     rintro _ ⟨⟩
 
+private theorem splitByLoop_append {r : α → α → Bool} {l g : List α} {a : α}
+    (h : (g.reverse ++ a :: l).Chain' fun x y ↦ r x y)
+    (ha : ∀ h : m ≠ [], r ((a :: l).getLast (cons_ne_nil a l)) (m.head h) = false) :
+    splitBy.loop r (l ++ m) a g [] = (g.reverse ++ a :: l) :: m.splitBy r := by
+  induction l generalizing a g with
+  | nil =>
+    rw [nil_append]
+    cases m with
+    | nil => simp [splitBy.loop]
+    | cons c m  =>
+      rw [splitBy.loop]
+      have := ha (cons_ne_nil c m)
+      rw [getLast_singleton, head_cons] at this
+      rw [this, splitByLoop_eq_append [_], splitBy]
+      simp
+  | cons b l IH =>
+    rw [cons_append, splitBy.loop]
+    split
+    · rw [IH]
+      · simp
+      · simp [h]
+      · rwa [getLast_cons (cons_ne_nil b l)] at ha
+    · rw [chain'_append_cons_cons, ← Bool.ne_false_iff] at h
+      have := h.2.1
+      contradiction
+
+set_option linter.docPrime false in
+theorem splitBy_of_chain' {r : α → α → Bool} {l : List α} (hn : l ≠ [])
+    (h : l.Chain' fun x y ↦ r x y) : splitBy r l = [l] := by
+  cases l with
+  | nil => contradiction
+  | cons a l => rw [splitBy, ← append_nil l, splitByLoop_append] <;> simp [h]
+
+@[simp]
+theorem splitBy_singleton {r : α → α → Bool} (a : α) : splitBy r [a] = [[a]] := by
+  apply splitBy_of_chain' <;> simp
+
+theorem splitBy_append {r : α → α → Bool} {l : List α} (hn : l ≠ [])
+    (h : l.Chain' fun x y ↦ r x y) (ha : ∀ h : m ≠ [], r (l.getLast hn) (m.head h) = false) :
+    (l ++ m).splitBy r = l :: m.splitBy r := by
```

Comment:
```suggestion
    (hlm : ∀ h : m ≠ [], r (l.getLast hn) (m.head h) = false) :
    (l ++ m).splitBy r = l.splitBy r ++ m.splitBy r := by
```
and you can recover your original result using `splitBy_of_chain'`

Reply:
This will require a bunch of reproving, but I'll try.